miner/scroll_worker.go

```diff
@@ -520,15 +507,40 @@ func (w *worker) newWork(now time.Time, parentHash common.Hash, reorging bool, r
 		header.Nonce = types.BlockNonce{}
 	} else {
 		prepareStart := time.Now()
-		if err := w.engine.Prepare(w.chain, header); err != nil {
+		// Note: this call will set header.Time, among other fields.
+		if err := w.engine.Prepare(w.chain, header, nil); err != nil {
 			return fmt.Errorf("failed to prepare header for mining: %w", err)
 		}
 		prepareTimer.UpdateSince(prepareStart)
-	}
 
-	var nextL1MsgIndex uint64
-	if dbVal := rawdb.ReadFirstQueueIndexNotInL2Block(w.eth.ChainDb(), header.ParentHash); dbVal != nil {
-		nextL1MsgIndex = *dbVal
+		if w.chainConfig.IsEuclidV2(header.Time) && !w.chainConfig.IsEuclidV2(parent.Time()) {
+			// We found a potential EuclidV2 transition block.
+			// We need to make sure that all the L1 messages V1 are consumed before we activate EuclidV2,
+			// since we can only include MessageQueueV2 messages after EuclidV2.
+			l1MessagesV1 := rawdb.ReadL1MessagesV1From(w.eth.ChainDb(), nextL1MsgIndex, 1)
+			if len(l1MessagesV1) > 0 {
+				// Reset Extra (it was unset by SystemContract)
+				header.Extra = w.extra
+
+				// Backdate the block to the parent block's timestamp -> not yet EuclidV2
```

Comment:
Same question on putting same timestamp.

Reply:
Same time, as implemented by @jonastheis. I don't see any issue with this, both engines allow this. And this is an edge case that probably won't be triggered. but just in case.